### ISSUE TITLE
Add new renaming rule variables

### ISSUE THREAD

**Issue by mcpierce** [comixed/comixed#512]
Some new variables to add to the renaming rules set:

NAME|VALUE
----|-----
$ISSUE(n)|Issue number padded with leading 0s to n-characters
$PUBMONTH|The publication month as text
$PUBYEAR|The publication year
$FORMAT|The comic format (if set)

**Comment by bareheiny** [comment]
Can I also request imprint, and some way to force double digit use for $ISSUE (so "01", "02" etc.).

**Comment by theotocopulitos** [comment]
Or even triple or cuádrale!

Actually some number padding scheme would be perfect, something like #ISSUE(3), for example

**Comment by mcpierce** [comment]
@bareheiny @theotocopulitos Good idea! I've added this to the table above.

**Comment by jdinda70** [comment]
Could you please look at adding the following as well:

$ISSUECOUNT(n) (so filenames could include "#001 (of 006)" for example)
$MONTH (more flexible than just coverdate ... $MONTH and $MONTH_NAME could potentially replace $COVERDATE
$FORMAT

**Comment by mcpierce** [comment]
@jdinda70 The only one I think may be an issue (pardon the pun) is the one for issue count: how would that work if it's an ongoing series?

**Comment by jdinda70** [comment]
I think that's something else that needs to be worked on in the renaming
rules.  In the case that any of the variables are blank, it would be good
if the variable was removed from the output (and I guess the tricky part
would be removing any other formatting characters around the variables).
Comictagger and Comicrack both do this now.

In ComicTagger, the feature is called "Smart Text Cleanup" and it's listed
as "Experimental" but in my experience it works very well.

[image: image.png]

Maybe you could raid the Comictagger code for ideas as to how to do this?

So, specifically, if the "The file renaming rule" was set to:
/$PUBLISHER/$SERIES ($VOLUME)/$SERIES ($VOLUME) #$ISSUE (of $ISSUECOUNT)
($YEAR-$MONTH) ($FORMAT) (as an example.  I know $MONTH, $ISSUECOUNT, and
$FORMAT are not available yet).

If any of the values were blank, the characters "around" the blank values
should be removed from the "new filename" string before the renaming occurs.

So if volume was blank the output would look like this:
/$PUBLISHER/$SERIES/$SERIES #$ISSUE (of $ISSUECOUNT) ($YEAR-$MONTH)
($FORMAT) (The parenthesis around $VOLUME would be removed.  Currently,
blank $VOLUME data is replaced by "UNKNOWN" in comixed)

So to answer your original question, if this feature was in in place and if
issue count was blank (as in the case of ongoing series), the output would
look like this: /$PUBLISHER/$SERIES ($VOLUME)/$SERIES ($VOLUME) #$ISSUE
($YEAR-$MONTH) ($FORMAT).  However, for issuecount to work, people would
need to manually enter it.

**Comment by jdinda70** [comment]
This file seems to have all of the renaming/smart cleanup code:

https://github.com/comictagger/comictagger/blob/develop/comictaggerlib/filerenamer.py

if self.smart_cleanup:
# smart cleanup means we want to remove anything appended to token if it's
empty
# (e.g "#%issue%" or "v%volume%")
# (TODO: This could fail if there is more than one token appended together,
I guess)
text_list = text.split()
# special case for issuecount, remove preceding non-token word,
# as in "...(of %issuecount%)..."
if token == '%issuecount%':
for idx, word in enumerate(text_list):
if token in word and not isToken(text_list[idx - 1]):
text_list[idx - 1] = ""
text_list = [x for x in text_list if token not in x]
return " ".join(text_list)
else:
return text.replace(token, "")

**Comment by ajhorvath20** [comment]
Would it be possible to replace colon with space dash?

**Comment by mcpierce** [comment]
@ajhorvath20 We definitely would need to replace any colon with a usable character.

**Comment by mcpierce** [comment]
@ajhorvath20 @bareheiny I've got a first pass at this. I've dropped the $CURRENTNAME variable since it strikes me as not something we'd want to support.

I'm still interested in $ISSUECOUNT, but we need a way to determine the total count before adding it to the variables. One possibility I see is that, at least with [ComicVine's APIs](https://comicvine.gamespot.com/api/documentation#toc-0-42) the volume can specify that total number of issues. But, TBH, issue count only sounds worthwhile for a miniseries rather than for a series that was canceled or restarted. So let's discuss this variable more in a new feature request.
